src/main/java/com/epam/bookstoreservice/service/impl/BookstoreServiceImpl.java

```diff
@@ -144,29 +117,77 @@ public List<BookResponseDTO> getBooksByCategoryAndKeyWord(String category, Strin
 
+    private void addBookEntityTotalCount(BookRequestDTO bookRequestDTO, BookEntity bookEntityFindById) {
 
-    private List<BookResponseDTO> getBooksByCategoryAndKeyWordUtil(String category, String keyword) {
-        int bookId = -1;
-        try {
-            bookId = Integer.parseInt(keyword);
-        } catch (Exception ex) {
-            logger.info(ex.getMessage());
+        Integer totalCount = bookRequestDTO.getTotalCount() + bookEntityFindById.getTotalCount();
+
+        bookEntityFindById.setTotalCount(totalCount);
+    }
+
+    private BookEntity checkAndGetBookEntityFindByIdIfExist(Integer id) {
+        Optional<BookEntity> bookFindById = bookDao.findById(id);
+
+        return bookFindById.orElseThrow(BookNotFoundException::new);
+    }
+
+    private void checkInventory(Integer id, Integer salesNumber) {
+        BookEntity bookEntityFindById = checkAndGetBookEntityFindByIdIfExist(id);
+
+        int remainingInventory = bookEntityFindById.getTotalCount() - salesNumber;
+
+        if (remainingInventory < 0) {
+            throw new InsufficientInventoryException();
         }
-        List<BookEntity> bookEntityListFindByCategoryAndKeyword = bookDao.findByCategoryAndKeyword(category, bookId, keyword);
 
-        if (bookEntityListFindByCategoryAndKeyword.isEmpty()) {
-            throw new BookNotFoundException();
+    }
+
+    private BookResponseDTO sellBookByIdAndSalesNumber(Integer id, Integer salesNumber) {
+
+        BookEntity bookEntityFindById = checkAndGetBookEntityFindByIdIfExist(id);
+
+        bookEntityFindById.setSold(bookEntityFindById.getSold() + salesNumber);
+        bookEntityFindById.setTotalCount(bookEntityFindById.getTotalCount() - salesNumber);
+
+        return bookDtoAndBookEntityMapper.entityToResponseDto(bookDao.save(bookEntityFindById));
+    }
+
+    private void validId(Integer id, BookRequestDTO bookRequestDTO) {
+
+        boolean idNonNullAndUnmatched
+                = Objects.nonNull(bookRequestDTO.getId()) && !Objects.equals(id, bookRequestDTO.getId());
+
+        if (idNonNullAndUnmatched){
+            throw new UnmatchedIdException();
         }
+    }
+
+    private List<BookResponseDTO> getBooksByCategoryAndKeyWordUtil(String category, String keyword) {
+
+        List<BookEntity> bookEntityListFindByCategoryAndKeyword = bookDao
+                .findByCategoryAndKeyword(category, getBookIdWhenKeywordIsNumber(keyword), keyword);
+
+        bookEntityListFindByCategoryAndKeyword
+                .forEach(bookEntity -> checkAndGetBookEntityFindByIdIfExist(bookEntity.getId()));
 
         return bookEntityListFindByCategoryAndKeyword
                 .stream()
-                .map(bookRequestDtoToBookEntity::entityToResponseDto).collect(Collectors.toList());
+                .map(bookDtoAndBookEntityMapper::entityToResponseDto).collect(Collectors.toList());
+    }
+
+
+    private int getBookIdWhenKeywordIsNumber(String keyword) {
+        int bookId = DEFAULT_BOOK_ID;
+        try {
```

Comment:
You can try another way to check if the keyword is Integer.
Hint: Character.isDigit(xx)